### DISCUSSION TITLE
hostname with multiple A records and one matches the local IP

### DISCUSSION THREAD

**Discussion by ech68** [ipxe/ipxe#651] [Q&A]
I've found empirically that if I have a hostname with multiple A records, one of which has the same IP as the host I'm booting in iPXE, that iPXE appears to discard all the other addresses, which then causes any connection to that host (like "chain http://webhost/foo") to timeout and never gets to attempt the other A records for the hostname. (if I put a loop to retry 10 times, it will fail every time)

I built a test version of iPXE (to enable ping and nslookup) and found that when I'm on a host in this situation... if I remove the A record with the local IP from that hostname, I can then "ping" and "chain" to it w/o problems.

It seems like this is actually exactly the opposite of what would make the most sense when faced with this setup - throwing away the local IP from the ones to use if it's part of a multi-A-record hostname while in iPXE would seem reasonable, since any service you're trying to connect to at the local IP isn't going to work.

I haven't been able to dig through and find a fix at the moment, but wanted to see if someone might know where the problem code that's always selecting the local IP from such a list of A records might be.

Thanks

**Comment by NiKiZe** [comment]
I would assume that building with DEBUG=dns would show you what is going on.

**Reply by ech68** [reply]
Your response got me thinking some more... because I'd pretty much already seen that nslookup in iPXE was always giving me back the local IP address when looking up that hostname with the local IP in the set of A records... so I tried doing the same lookup after getting the host booted up and running linux to see if the same thing happened.

Sure enough, every time I look up that hostname from a host that has the local/requesting IP in the list of A records, the first entry returned is always the requesting IP, so it appears to be something configured on our DNS servers.

I wonder if there's a way to tell iPXE to randomize the address it tries to connect to in that list, rather than always using the first? Alternately, it would be nice if iPXE would try all addresses returned instead of failing to connect when the first address can't be reached (it gets connection timeout errors when trying to chain load from itself).

**Reply by NiKiZe** [reply]
If you have multiple records it is "random" but do use tcpdump or similar to see the queries.

It would indeed be good to try other hosts on connection failures, if it isn't already.

**Reply by ech68** [reply]
Unfortunately is isn't random - every time I query for this hostname from a host with its IP in the set of A records, that IP is always returned first, and it looks like iPXE only attempts to connect to the first A record returned.

Enabling dns debugging only ever shows one address resolved even when there are multiple A records for the host, and for other multi-a-record hosts where the IPs are all foreign to this host, I do get a random IP address shown by the dns debugging in iPXE on subsequent lookups (with ping enabled, I can ping repeatedly and see the change).

**Reply by NiKiZe** [reply]
Usually DNS servers use round robin (random) but they could be ordered. Check your DNS servers configuration for the records and/or provide details on which server you are using to maybe get further assistance.

As I understand your setup, you have both local and public IPs in DNS? If so that is a broken configuration.

**Reply by stappersg** [reply]
The problem could indeed be caused by **several truths** in DNS.

And it is wrong to have a _server in DNS_ that is _in maintiance mode_. 

> multiple A records, one of which has the same IP as the host I'm booting in iPXE,


Go for a truth that is truly true.

**Comment by ech68** [comment]
The DNS server is only internal (no public and local servers) and is a Windows Domain Controller which maybe should/could be configured differently- I mentioned "foreign" IP addresses as those that aren't the IP address of the host making the queries.

However, other dns clients, like "curl" as a simple example, does try connecting to all IP addresses even if the first one in the list fails, and that's the problem I see with the iPXE dns resolution - it only tries the firs IP it gets back and if that fails to connect, decides that it's done and errors out.

Regarding "it's wrong to have a server in DNS that is in maintenance mode" - are you saying that I should remove the server from the list of IPs ever time it reboots or happens to crash? I'd counter that since other (many/most) client apps that use DNS _do_ use all the addresses and don't just stop at trying to connect to the first one, that that's a problem with iPXE's behavior wrt dns and its handling of the addresses it gets back.

**Reply by NiKiZe** [reply]
AFAIK there is no RFC that requires a client to try multiple records from DNS, that is optional. I again agree that it would be a nice feature.

**Comment by stappersg** [comment]
Summary in my words:
* There is `curl`
* There is `ipxe`
* `curl` is a client
* `ipxe` is a client
* `curl` is a HTTP client
* `ipxe` is a HTTP client
* `curl` is  a HTTP client and also a DNS client
* `ipxe` is a HTTP client and also a DNS client
* `curl` behaves in a certian way
* `ipxe` is expected by @ech68 to behave like `curl`
* `curl` does nice things
* `ipxe` does other nice things
